### ISSUE TITLE
404 With Multiple Images in One Request

### ISSUE THREAD

**Issue by jshlfts32** [expressjs/multer#582]
I’m receiving a 404 Not Found error when I attempt to submit a form that has two images uploaded. Neither image is a required field, so when I upload one image (can be either) everything works correctly, and I do not get an error. The issue only occurs when I upload both in one request. Checking my uploads folder, when I get a 404, both images are correctly uploaded.
Here is the code for multer and resizing:
```javascript
const multerOptions = {
  storage: multer.memoryStorage(),
  fileFilter(req, file, next) {
    const isPhoto = file.mimetype.startsWith('image/');
    if (isPhoto) {
      next(null, true);
    } else {
      next({ message: 'That filetype isn\'t allowed!' }, false);
    }
  },
};

export const upload = multer(multerOptions).fields([
  { name: 'avatar' },
  { name: 'accolade' },
]);

export const resize = async (req, res, next) => {
  if (!req.files) {
    next();
    return;
  }
  Object.keys(req.files).forEach(async (file) => {
    const extension = req.files[file][0].mimetype.split('/')[1];
    req.body[file] = `${uuid.v4()}.${extension}`;
    const uploaded = await jimp.read(req.files[file][0].buffer);
    if (file === 'avatar') {
      await uploaded.resize(300, jimp.AUTO);
    } else if (file === 'accolade') {
      await uploaded.resize(30, jimp.AUTO);
    }
    await uploaded.write(`./public/uploads/${req.body[file]}`);
    next();
  });
};
```

I’m passing both functions to this route:
```javascript
router.post(
  '/team-members/add/:id',
  authController.authCheck,
  userController.isAdmin,
  userController.upload,
  userController.resize,
  userController.validateUser,
  catchErrors(userController.addTeamMember),
);
```

The final method, `addTeamMember` fires correctly and the user is added, but I get a 404 not found and `Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client`. Any ideas?

**Comment by jshlfts32** [comment]
Hi Linus, thanks for the response. I'm sure you're right and I'm just missing it. 

Here's the code for each of the middlewares:

```javascript
export const authCheck = (req, res, next) => {
  (req.isAuthenticated()) ? next() : res.redirect('/login');
};
```
```javascript
export const isAdmin = (req, res, next) => {
  (req.user.role !== 'admin') ? res.redirect('/dashboard') : next();
};
```
(upload and resize listed in op)

```javascript
export const validateUser = (req, res, next) => {
  req.checkBody('firstName', 'There must be a first name!').notEmpty();
  req.checkBody('lastName', 'There must be a last name!').notEmpty();
  req.checkBody('email', 'There must be an email!').notEmpty();
  req.checkBody('role', 'A role must be specified!').notEmpty();

  const errors = req.validationErrors();
  if (errors) {
    req.flash('error', errors.map(err => err.msg));
    res.redirect('back');
  }
  next();
};
```

And the final method (wrapped in a catcher function):
```javascript
export const addTeamMember = async (req, res) => {
  const org = await Org.findOne({ _id: req.params.id });
  if (org) {
    const newUser = await (new User(req.body));
    newUser.organization = org._id;
    newUser.invitation = true;
    await newUser.save();
    await org.update({ $push: { users: newUser } });

    const inviteLink = `http://${req.headers.host}/join/${org._id}`;
    await send({
      user: newUser,
      filename: 'invitation',
      subject: `Welcome ${newUser.email}`,
      inviteLink,
    });
    req.flash('success', `Yay! An invitation has been sent to ${newUser.email}`);
    res.redirect(`/team-members/${org._id}`);
  } else {
    req.flash('error', 'No organization found!');
    req.redirect('back');
  }
};
```

I only get the error when I upload both an avatar and an accolade in one request. If I upload just one in a single request, I get no errors. In both cases, the image(s) are uploaded to the uploads directory I've specified, the user is added to my db, and an email for an invite is fired off. The redirect on success is a single get request to a view with the same authCheck and isAdmin middlewares.

**Comment by LinusU** [comment]
It could be this?

```diff
 export const validateUser = (req, res, next) => {
   req.checkBody('firstName', 'There must be a first name!').notEmpty();
   req.checkBody('lastName', 'There must be a last name!').notEmpty();
   req.checkBody('email', 'There must be an email!').notEmpty();
   req.checkBody('role', 'A role must be specified!').notEmpty();
 
   const errors = req.validationErrors();
   if (errors) {
     req.flash('error', errors.map(err => err.msg));
     res.redirect('back');
+    return
   }
   next();
 };
```

**Comment by jshlfts32** [comment]
Thanks again for responding! 

I added the return statement, but unfortunately still getting the error when uploading two images at the same time.

I've added a question at SO, and will close this issue as it seems it probably isn't an issue with this package from your responses. 
https://stackoverflow.com/questions/49982952/nodejs-multer-receiving-404-and-cannot-set-headers-when-uploading-multiple-fil
